### ISSUE TITLE
the setting and application of the parameter "target-mu"

### ISSUE THREAD

**Issue by moyulyy** [shankar1729/jdftx#214]
Hello, Shankar,

I am a JDFTX beginner, thank you very much for your issue #211,issue #212 solutions, I learned a lot.

In the subsequent study, I have some questions about the setting and application of the parameter "target-mu", and I am looking forward to your help again.

**1. If the value of "target-mu" is not set, whether JDFTX performs the calculation process of constant charge model(CCM)**.
Here's an example,
#-------- Electronic minimization -----------
#	FillingsUpdate:  mu: -0.181866701  nElectrons: 336.000000  magneticMoment: [ Abs: 0.00014  Tot: +0.00000 ]
#ElecMinimize: Iter:   0  F: -635.576212354643189  |grad|_K:  3.453e-07  alpha:  1.000e+00
#	FillingsUpdate:  mu: -0.182047753  nElectrons: 336.000000  magneticMoment: [ Abs: 0.00014  Tot: +0.00000 ]
#	SubspaceRotationAdjust: set factor to 0.309
#ElecMinimize: Iter:   1  F: -635.576236451848217  |grad|_K:  1.430e-07  alpha:  6.485e-01  linmin:  1.345e-07  t[s]:  54810.84
#	FillingsUpdate:  mu: -0.181811853  nElectrons: 336.000000  magneticMoment: [ Abs: 0.00013  Tot: +0.00000 ]
#	SubspaceRotationAdjust: set factor to 0.222

In the process of vacuum structure optimization,Note that the nElectrons reported by each ElecMinimize step in the output report are identical to some value (336.000000) when the "target-mu" parameter is not set.Does this mean that the calculation is a constant charge model(CCM) with a constant number of electrons? Essentially, the number of electrons reported by the nElectrons doesn't change, so does that mean the system's charge doesn't change during the calculation?

Also, consult you for the exact meaning of "Abs" and "Tot" in the magneticMoment parameters reported by each ElecMinimize calculation step.

**2. How do I set the initial value of "target-mu" parameter.**

I have optimized the slab structure by vacuuming it without setting the "target-mu" parameter,and obtained the value of the last mu (namely -0.181933018 below), namely the PZC of this slab.

#-------- Electronic minimization -----------
#	FillingsUpdate:  mu: -0.181808059  nElectrons: 336.000000  magneticMoment: [ Abs: 0.00009  Tot: +0.00000 ]
#
#ElecMinimize: Iter:   4  F: -635.576242956786245  |grad|_K:  6.685e-08  alpha:  6.405e-01  linmin:  4.201e-08  t[s]:  27606.98
#	FillingsUpdate:  mu: -0.181933018  nElectrons: 336.000000  magneticMoment: [ Abs: 0.00009  Tot: +0.00000 ]
#	SubspaceRotationAdjust: set factor to 0.329
#ElecMinimize: Iter:   5  F: -635.576243675123237  |grad|_K:  4.547e-08  alpha:  5.159e-01  linmin: -2.362e-07  t[s]:  27682.42
#ElecMinimize: Converged (|Delta F|<1.000000e-06 for 2 iters).

Next, I was confused about how to set the initial value of the "target-mu" parameter.

A clear strategy is that I must start with a neutral calculation in a non-adsorptive NaF electrolyte. However, when I calculate my system in imitation of the "fixed-potential calculations" tutorial, the calculation results show that the calculation process of "IonicMinimize" does not converge, as shown below:

#IonicMinimize: Iter:   0  G: -256.391511910029692  |grad|_K:  1.313e+02  t[s]:    569.93
#IonicMinimize: None of the convergence criteria satisfied after 0 iterations.

The results suggest that I should set the parameter "ionic-minimize" for structural optimization calculation.

However, should I discard the previous result of structural optimization without setting "target-mu" parameter in vacuum and use a new structure optimization calculation in non-adsorbable NaF electrolyte to get the PZC of this slab again?
Alternatively, based on the previous result of structural optimization without setting "target-mu" parameter in vacuum, the structural optimization calculation in non-adsorbable NaF electrolyte was carried out to obtain the PZC of this slab again.

In fact, I puzzled before the necessity of structural optimization by vacuum without setting the "target-mu" parameter. (For aqueous electrochemical systems)

In addition, in the description of "target-mu" in the tutorial, the parameter "mu" is defined as **mu = -(Vref + V)/27.2114,** which is the same formula applied to calibrate the Standard Hydrogen Electrode with CANDLE model discussed in Issue #17.

However, in the tutorial "fixed-potential calculations", the calculation method of parameter "mu" seems to be: **mu = -PZC + V/27.2114**, as shown below.

# We will now run a sequence of fixed-potential calculations for mu varying from -1 to +1 eV relative to the neutral potential (PZC) of 
# -0.2015 Hartrees (final mu from Neutral.out).
and
# export mu="$(echo $iMu | awk '{printf("%.4f", -0.2015+0.1*$1/27.2114)}')"

Also, in the tutorial "Two-dimensional Materials",Seems to suggest that mu(from final FillingsUpdate line) = PZC = Fermi level,as shown below.
#Edit the bandstructure plot to reference the enegies against the Fermi level (mu from final FillingsUpdate line) and convert it to eV

Then the calculation method of parameter "mu" can be: **mu = -Efermi_level + V/27.2114**.

I am very troubled about the accurate calculation formula of external electric potential V applied after using CANDLE model to calibrate standard hydrogen electrode.Because there are three parameters, Fermi Level, CANDLE model calibration standard hydrogen electrode, and external electric potential V applied, any of the above calculation formula obviously involves only two parameters.

Finally, I question the basis on which I chose to apply V, or in what way I should determine the value of V applied. Is it about Fermi levels, the number of charges, multiple sets of calculations, or is it about empirical selection?

Thank you again for everything you've done.

**Comment by shankar1729** [comment]
1. Yes, the default mode of JDFTx is standard constant-charge DFT calculations. (JDFTx is a general-purpose DFT code, not just for electrochemistry.) For magnetization, Abs is integral of |M(r)|on the unit cell, where M(r) is the magnetization density, while tot is integral of M(r) (without abs). So, for an antiferromagnetic system for example, Abs will be non-zero while Tot may be zero.

2. I'm not clear on what you are asking here. Could you please summarize in a couple of sentences what the issue is? When you got the IonicMinimize with the really large force, what mu did you use? Also, the tutorials lay out the procedure you would typically adopt. Do a neutral calculation first, followed by the one at mu. If you want to optimize geometry, you should typically optimize the neutral geometry as well. 

Finally, you may want to read the GC-DFT and cited solvation papers in more detail to understand mu and referencing. You may be mixing up experimental electrode reference potential calibration vs. results for V w.r.t PZC (i.e. neutral). As far as JDFTx is concerned, mu is always electron chemical potential in absolute Hartree units (both in input to target-mu and as reported in FillingsUpdate). What values you choose for mu and how you connect it to experimental electrode potentials in Volts is a separate question, for which you can adopt various schemes discussed in the literature. (The tutorials describe one such scheme.)

Best,
Shankar

**Comment by moyulyy** [comment]
Dear Shankar
Thank you very much for your answer, which has solved my confusion and benefited me a lot. 

I apologize for not being clear about how to set the "target-mu" parameter.
But I do know that the correct calculation is to do the neutral calculation first, and then do the calculation at "mu".I'm going to try to do it this way.

Next, I will read more literature about GC-DFT and solvation, learn information about "mu" and the method of setting "mu" value.

Thank you again for your help.

**Comment by moyulyy** [comment]
Dear Shankar,

Sorry to bother you again, I have a stupid question, but I want to confirm it to make sure the calculation is correct.

**Whether a continuous solvation model of the electrolyte (solvent + ion) must be used when I use the parameter "target-mu" for my calculations. 
In other words, can I use JDFTX to calculate the constant-potential model in vacuum**?

Although in the tutorial and the paper "JDFTx: Software for Joint Density functional Theory " the constant-potential calculation of continuous solvation models using electrolytes (solvents + ions) is shown.

However, I reported an output error when I tried to calculate the constant-potential model in vacuum using the parameter "target-mu".

The report says,
#--------------------------------------------------------------------------
#Encountered 1 errors while parsing input:
#Command 'target-mu' is missing dependencies { fluid-anion fluid-cation }
#Input parsing failed with 1 errors (run with -t for command syntax)
#--------------------------------------------------------------------------

Clearly, I need to add a continuous solvation model using an electrolyte (solvent + ion).

However, I would like to have your confirmation so as to avoid being misled by possible mistakes.


Thank you very much for your efforts, which helped me a lot.

**Comment by shankar1729** [comment]
Yes, that is right: you need a solvation model with ions to provide the countercharge as the electronic system gets charged. 

Without an electrolyte solvation, the electrostatic potential of a charged electronic system does not decay exponentially away from the system. Correspondingly, mu cannot be calibrated against a reference far away, and the whole idea of fixed-potential calculations becomes meaningless. This is discussed in detail in the GC-DFT (2017 JCP) as well as the solvation model papers cited therein. In particular, read the 2012 PRB by Letchworth-Weaver and Arias for a long explanation of these points.

Best,
Shankar
